thanosconvert: convert block meta from Thanos to Cortex

milesbxf opened the pull request:
**What this PR does**:

thanosconvert is a small tool to iterate over a bucket and convert block-level metadata (in meta.json) so that it is readable by Cortex. This tool assumes that users have already copied Thanos blocks over to a destination bucket, with blocks organised by user at the top level.

This is a relatively simple serial implementation - it's pretty quick to run so making it concurrent is probably unnecessary, at least for a v1. I've only tested this with S3, but since it uses `github.com/thanos-io/thanos/pkg/objstore` it should work with all the other supported block stores. It reuses the same `pkg/storage/bucket.Config` struct for bucket configuration so should be easy to configure for users who already have the store-gateway/querier running.

It's idempotent so is safe to run multiple times against the same bucket, and leaves Cortex blocks untouched. It will log any errors it encounters but keeps going.

I'm happy that this is mostly complete, but there's a few things left to do - I'd love some feedback before I finish it off!

**Which issue(s) this PR fixes**:
Fixes #2920

**Checklist**
- [x] Add CLI usage message
- [x] Clean up, comment public API methods
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pracucci commented on cmd/thanosconvert/main.go:
```diff
@@ -0,0 +1,78 @@
+	iterCtx := context.Background()
+	results, err := converter.Run(iterCtx)
+	if err != nil {
+		level.Error(util.Logger).Log("msg", "iterate blocks", "err", err)
```
```suggestion
		level.Error(util.Logger).Log("msg", "error while iterating blocks", "err", err)
```

pracucci commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,195 @@
+type Results map[string]PerUserResults
+
+// NewThanosBlockConverter creates a ThanosBlockConverter
+func NewThanosBlockConverter(ctx context.Context, cfg bucket.Config, dryRun bool, logger log.Logger, reg prometheus.Registerer) (*ThanosBlockConverter, error) {
```
I would remove `reg prometheus.Registerer` given this tool doesn't export metrics anyway, and just pass `nil` as registered to `bucket.NewClient()` (when `nil` metrics registration will be skipped).

pracucci commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,195 @@
+	results := make(Results)
+
+	// discover users in the bucket
+	var users []string
+	err := c.bkt.Iter(ctx, "", func(o string) error {
+		users = append(users, strings.TrimSuffix(o, "/"))
+		return nil
+	})
```
You may want to reuse the scanner we already have:

```
scanner := cortex_tsdb.NewUsersScanner(c.bkt, cortex_tsdb.AllUsers)
users, _, err := scanner.ScanUsers(ctx)
```

pracucci commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,195 @@
+func (c ThanosBlockConverter) convertUser(ctx context.Context, user string) (PerUserResults, error) {
+	results := PerUserResults{}
+
+	err := c.bkt.Iter(ctx, user, func(o string) error {
```
I would wrap `c.bkt` with `bucket.NewUserBucketClient()`. This will mask you the first object prefix segment (which is the `<tenant ID>/`).

pracucci commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,195 @@
+	err := c.bkt.Iter(ctx, user, func(o string) error {
+		// get block ID from path
+
+		parts := strings.Split(o, "/")
```
Once you use the `UserBuckerClient`, this should just be a matter of:

```
o = strings.TrimSuffix(o, "/")
blockID, err := ulid.Parse(o)
```

If the parse fail, just skip it (no error) because it means it's not a block. We don't have just blocks in bucket.

pracucci commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,195 @@
+			return nil
+		}
+
+		meta := &metadata.Meta{}
+		decoder := json.NewDecoder(r)
+		err = decoder.Decode(meta)
```
Please use `metadata.Read()` from Thanos package `github.com/thanos-io/thanos/pkg/block/metadata/`

pracucci commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,195 @@
+		// don't need to add to changesRequired, since the code below will notice lack of org id
+	}
+
+	org, ok := meta.Thanos.Labels["__org_id__"]
```
Instead of the hardcoded string `"__org_id__"`, please use `cortex_tsdb.TenantIDExternalLabel`.

pracucci commented on tools/thanosconvert/thanosconvert_test.go:
```diff
@@ -0,0 +1,355 @@
+	}
+}
+
+func CortexMeta(user string) metadata.Meta {
```
No need to export this function.

pracucci commented on tools/thanosconvert/thanosconvert_test.go:
```diff
@@ -0,0 +1,355 @@
+	}
+}
+
+func ThanosMeta() metadata.Meta {
```
No need to export this function.

pracucci commented on tools/thanosconvert/thanosconvert_test.go:
```diff
@@ -0,0 +1,355 @@
+	}
+}
+
+func GetLogger() log.Logger {
```
No need to export this function.

pracucci commented on tools/thanosconvert/thanosconvert_test.go:
```diff
@@ -0,0 +1,355 @@
+	return logger
+}
+
+func MockBucket(data fakeBucket) *bucket.ClientMock {
```
No need to export this function.

pracucci commented on tools/thanosconvert/thanosconvert_test.go:
```diff
@@ -0,0 +1,355 @@
+			name: "bucket fully converted is a noop",
+			bucketData: fakeBucket{
+				"user1": map[string]metadata.Meta{
+					"block1": CortexMeta("user1"),
```
Once you start using `ulid.Parse()` these tests will fail because `"block1"` is not a valid block ID. You could just define at the beginning of the test function something like the following, and then use such variables across the test cases:

```
block1 := ulid.MustNew(1, nil)
block2 := ulid.MustNew(2, nil)
```

pracucci reviewed (commented):
Fantastic job! We wanted this since a long time, thanks! I left few comments here and there, but mostly nits that should be very quick to address.

Other than the comments:
- Could you add it to the list of experimental features listed in `docs/configuration/v1-guarantees.md`, please?
- Could you mention the tool in `docs/blocks-storage/migrate-storage-from-thanos-and-prometheus.md`? There's a sentence like "no tool is provided to migrate TSDB blocks from Thanos" while now there is 😉

milesbxf commented:
> Fantastic job! We wanted this since a long time, thanks! I left few comments here and there, but mostly nits that should be very quick to address.
> 
> Other than the comments:
> 
> * Could you add it to the list of experimental features listed in `docs/configuration/v1-guarantees.md`, please?
> * Could you mention the tool in `docs/blocks-storage/migrate-storage-from-thanos-and-prometheus.md`? There's a sentence like "no tool is provided to migrate TSDB blocks from Thanos" while now there is 😉

@pracucci thanks for the detailed review! I think I've addressed all your comments - I've added a new commit to make it easier to review the changes, but can squash it down once you're happy. I'll make the docs changes today.

milesbxf commented:
👋  @pracucci I've updated the docs and added a CHANGELOG entry. Is there anything else I need to do?

pracucci commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,188 @@
+
+		// retrieve meta.json
+
+		metaKey := fmt.Sprintf("%s/%s/meta.json", user, blockID)
+		r, err := c.bkt.Get(ctx, metaKey)
```
Use the `userBucketClient` please. It means the path to `meta.json` will be:
`metaKey := fmt.Sprintf("%s/meta.json", blockID)`

pstibrany replied:
After using `userBucketClient`, we can also use `block.DownloadMeta()` which will download and parse the meta.json file (it doesn't check for Cortex orgid label, so that's fine), to simplify the code here.

pracucci commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,188 @@
+	var changesRequired []string
+
+	if meta.Thanos.Labels == nil {
+		// TODO: no "thanos" entry could mean this block is corrupt - should we bail out instead?
```
I would not bail out. If we don't require the Thanos entry, the tool should work with Prometheus blocks too, which I believe is a plus for the community.

pracucci reviewed (approved):
Good job, thanks! I left a couple of nits. Can you confirm you manually tested it after changes? I haven't got the time to manually test it.

@pstibrany could you do the 2nd review, please?

pstibrany commented on docs/blocks-storage/migrate-storage-from-thanos-and-prometheus.md:
```diff
@@ -18,18 +18,57 @@ The Cortex blocks storage has few requirements that should be considered when mi
+
+`thanosconvert` can iterate over a Cortex bucket and make sure that each `meta.json` has the correct `thanos` > `labels` layout.
+
+⚠️  `thanosconvert` will modify files in the bucket you specify. It's recommended that you have backups or enable object versioning before running this tool.
```
```suggestion
⚠ Warning ⚠ `thanosconvert` will modify files in the bucket you specify. It's recommended that you have backups or enable object versioning before running this tool.
```

pstibrany commented on docs/blocks-storage/migrate-storage-from-thanos-and-prometheus.md:
```diff
@@ -18,18 +18,57 @@ The Cortex blocks storage has few requirements that should be considered when mi
+⚠️  `thanosconvert` will modify files in the bucket you specify. It's recommended that you have backups or enable object versioning before running this tool.
 
-For each block copied/uploaded to the Cortex bucket, the `meta.json` should be manipulated. The easiest approach would be iterating the tenants and blocks in the bucket and for each block:
+To run `thanosconvert`, you need to provide a YAML file with bucket configuration in the same format as the [blocks storage bucket configuration](../configuration/config-file-reference.md#blocks_storage_config). For example, for s3:
```
```suggestion
To run `thanosconvert`, you need to provide it with the bucket configuration in the same format as the [blocks storage bucket configuration](../configuration/config-file-reference.md#blocks_storage_config).
```

(command line options work too here)

pstibrany commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,188 @@
+
+func NewPerUserResults() PerUserResults {
+	return PerUserResults{
+		FailedBlocks:    []string{},
```
Nit: this creates 0-sized slice, which is unnecessary. As it is now, `PerUserResults{}` is perfectly fine default value, and doesn't even need "New" function -- which isn't used anywhere.

(Furthermore, since methods are defined on `*PerUserResults`, this method should return pointer. But I'd suggest to remove it completely.)

milesbxf replied:
Turns out I wasn't even using this constructor anyway 😄

pstibrany commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,188 @@
+	scanner := cortex_tsdb.NewUsersScanner(c.bkt, cortex_tsdb.AllUsers, c.logger)
+	users, _, err := scanner.ScanUsers(ctx)
+	if err != nil {
+		return results, errors.Wrap(err, "error while scanning users")
```
Nit: unify "users" vs "tenants" terminology in the log and error messages.

milesbxf replied:
I've gone with "users" - which one is more correct? Is there a distinction?

pstibrany replied:
There is no distinction in Cortex. Our codebase uses both terms interchangeably, and it is sometimes confusing for people.

pstibrany commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,188 @@
+	userBucketClient := bucket.NewUserBucketClient(user, c.bkt)
+	err := userBucketClient.Iter(ctx, "", func(o string) error {
+		// get block ID from path
+		o = strings.TrimSuffix(o, "/")
```
We can shorten the code here by using `id, ok := block.IsBlockDir(o)`.

pstibrany commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,188 @@
+
+		if len(changesRequired) > 0 {
+			if c.dryRun {
+				level.Debug(c.logger).Log("msg", "Block requires changes, not uploading due to dry run", "block", blockID, "changes_required", strings.Join(changesRequired, ","))
```
Do we need to repeat "not uploading due to dry run" message for each block? Perhaps just short "(dry-run)" would do.

pstibrany commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+func (c ThanosBlockConverter) convertUser(ctx context.Context, user string) (PerUserResults, error) {
+	results := PerUserResults{}
```
I'd suggest to simplify this code by not keeping `PerUserResults` at all. The tool could simply use logging to log converted blocks, progress and failures, and not retain anything in the memory after processing of the user is done. It's only printed anyway -- we could print it immediately.

milesbxf replied:
Yes, good point - in my head I was thinking of putting out more sophisticated reporting (perhaps saving to a file), but this is probably unnecessary. On the other hand, this makes testing significantly easier, as we can run assertions on the contents of the results

pstibrany replied:
Good point about testing.

pstibrany commented on cmd/thanosconvert/main.go:
```diff
@@ -0,0 +1,88 @@
+	logfmt, loglvl := logging.Format{}, logging.Level{}
+	logfmt.RegisterFlags(flag.CommandLine)
+	loglvl.RegisterFlags(flag.CommandLine)
+	flag.StringVar(&configFilename, "config", "", "Path to bucket config YAML")
```
Nit: I'd suggest supporting command line options too (or perhaps as the only way of configuring). Bucket config needs just a couple of options.

pstibrany commented on cmd/thanosconvert/main.go:
```diff
@@ -0,0 +1,88 @@
+
+var (
+	configFilename string
+	dryRun         bool
```
Nit: no need for global variables, move to main() method.

pstibrany reviewed (approved):
Thank you, this is useful tool. I left bunch of comments suggestions if you feel like working on this code bit more, but they are not blocking... the tool already looks useful as-is.

milesbxf commented:
> Thank you, this is useful tool. I left bunch of comments suggestions if you feel like working on this code bit more, but they are not blocking... the tool already looks useful as-is.

Thanks for the review - I figured I'd implement a few of your suggestions whilst the PR is still open 🙏

milesbxf commented:
> Good job, thanks! I left a couple of nits. Can you confirm you manually tested it after changes? I haven't got the time to manually test it.

Thanks! Just to confirm, I've manually tested this after the latest batch of changes (again, only with S3)

milesbxf commented:
Looks like an unrelated test failed:
```
--- FAIL: TestAlertmanager_ServeHTTP (0.24s)
    multitenant_test.go:90: 
        	Error Trace:	multitenant_test.go:90
        	            				testing.go:846
        	            				testing.go:876
        	            				testing.go:1048
        	Error:      	Received unexpected error:
        	            	unlinkat /tmp/alertmanager319454365: directory not empty
        	Test:       	TestAlertmanager_ServeHTTP
FAIL
coverage: 65.6% of statements in github.com/cortexproject/cortex/pkg/alertmanager
FAIL	github.com/cortexproject/cortex/pkg/alertmanager	32.268s
FAIL
```

I can report an issue for this if you think it's a flake - in the meantime how do I rerun CI?

pstibrany commented:
> in the meantime how do I rerun CI

There is "Rerun all jobs" button in top-right corner of the workflow run. I've restarted it now, just in case you don't have permissions and don't see that button.

pstibrany commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,167 @@
+
+		if len(changesRequired) > 0 {
+			if c.dryRun {
+				level.Info(c.logger).Log("msg", "Block requires changes (dry-run)", "block", blockID.String(), "changes_required", strings.Join(changesRequired, ","))
```
I'd suggest to add "user" into all log messages here. (Or use `ulogger := log.With(c.logger, "user", user)` use wrapped ulogger only in entire method.)

pstibrany commented on tools/thanosconvert/thanosconvert.go:
```diff
@@ -0,0 +1,167 @@
+func convertMetadata(meta metadata.Meta, expectedUser string) (metadata.Meta, []string) {
+	var changesRequired []string
+
+	if meta.Thanos.Labels == nil {
```
No need for this check or creating of empty `meta.Thanos.Labels`, since it will be replaced anyway. Reading from `nil` map, or deleting from `nil` map will work as expected. (I'd suggest adding unit test for that)

pstibrany commented on cmd/thanosconvert/main.go:
```diff
@@ -0,0 +1,89 @@
+	logger, err := log.NewPrometheusLogger(loglvl, logfmt)
+	if err != nil {
+		fmt.Printf("failed to create logger: %v\n", err)
+		flag.Usage()
```
Printing usage hides the error message. I'd suggest removing flag.Usage() calls after print errors.

pstibrany commented on cmd/thanosconvert/main.go:
```diff
@@ -0,0 +1,89 @@
+	if configFilename != "" {
+		buf, err := ioutil.ReadFile(configFilename)
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to load config file", "err", err, "filename", configFilename)
```
This methods uses three different ways to print errors: logger, `fmt.Printf` (stdout) and `fmt.Fprintf(flag.Commandline.Output())` which uses stderr. I'd suggest to print errors followed by os.Exit(1) to stderr.

pstibrany commented on cmd/thanosconvert/main.go:
```diff
@@ -0,0 +1,89 @@
+		os.Exit(1)
+	}
+
+	fmt.Println("Results:")
```
We use logger for logging the progress in `Run` method. Perhaps we should use logger for logging the result too?

pstibrany commented on cmd/thanosconvert/main.go:
```diff
@@ -0,0 +1,89 @@
+
+	iterCtx := context.Background()
+	results, err := converter.Run(iterCtx)
+	if err != nil {
```
There may be some results already when this error happens. Should this code simply log results first, and error (if any) as last?

pstibrany reviewed (approved):
Thanks for addressing my feedback. I've left few more (non-blocking) comments to take a look at, if you want.

milesbxf commented:
> Thanks for addressing my feedback. I've left few more (non-blocking) comments to take a look at, if you want.

Thanks @pstibrany - I've addressed some of these 🙏 I probably won't be able to spend much more time on this, so let me know if there's anything blocking the merge you'd like me to address

pracucci commented:
Thanks a lot @milesbxf for your hard work!

> I probably won't be able to spend much more time on this, so let me know if there's anything blocking the merge you'd like me to address

Sure. It's a great tool, let's ship it! 🚀
